libvast/vast/concept/convertible/data.hpp

```diff
@@ -0,0 +1,439 @@
+template <class Type, class T>
+  requires(!detail::integral<T>)
+|| detail::same_as<bool, T> caf::error
+  convert(const T& src, T& dst, const Type&) {
+  dst = src;
+  return caf::none;
+}
+
+// Dispatch to standard conversion.
+// clang-format off
+template <class From, class To, class Type>
+  requires (!detail::same_as<From, To>) &&
+           detail::convertible_to<From, To>
+caf::error convert(const From& src, To& dst, const Type&) {
+  dst = src;
+  return caf::none;
+}
+// clang-format on
+
+// Overload for counts.
+template <detail::unsigned_integral To>
+caf::error convert(const count& src, To& dst, const count_type&) {
+  if constexpr (sizeof(To) >= sizeof(count)) {
+    dst = src;
+  } else {
+    if (src < std::numeric_limits<To>::min()
+        || src > std::numeric_limits<To>::max())
+      return caf::make_error(ec::convert_error,
+                             fmt::format(": {} can not be represented by the "
+                                         "target variable [{}, {}]",
+                                         src, std::numeric_limits<To>::min(),
+                                         std::numeric_limits<To>::max()));
+    dst = detail::narrow_cast<To>(src);
+  }
+  return caf::none;
+}
+
+// Overload for integers.
+template <detail::signed_integral To>
+caf::error convert(const integer& src, To& dst, const integer_type&) {
+  if constexpr (sizeof(To) >= sizeof(integer::value)) {
+    dst = src.value;
+  } else {
+    if (src.value < std::numeric_limits<To>::min()
+        || src.value > std::numeric_limits<To>::max())
+      return caf::make_error(ec::convert_error,
+                             fmt::format(": {} can not be represented by the "
+                                         "target variable [{}, {}]",
+                                         src, std::numeric_limits<To>::min(),
+                                         std::numeric_limits<To>::max()));
+    dst = detail::narrow_cast<To>(src.value);
+  }
+  return caf::none;
+}
+
+// Overload for enums.
+// clang-format off
+template <class To>
+  requires std::is_enum_v<To>
+caf::error convert(const std::string& src, To& dst, const enumeration_type& t) {
+  auto i = std::find(t.fields.begin(), t.fields.end(), src);
+  if (i == t.fields.end())
+    return caf::make_error(ec::convert_error, ": {} is not a value of {}", src,
+        detail::pretty_type_name(dst));
+  dst = detail::narrow_cast<To>(std::distance(t.fields.begin(), i));
+  return caf::none;
+}
+// clang-format on
+
+// Overload for lists.
+template <detail::has_push_back To>
+caf::error convert(const list& src, To& dst, const list_type& t) {
+  size_t num = 0;
+  for (const auto& element : src) {
+    typename To::value_type v{};
+    if (auto err = convert(element, v, t.value_type))
+      return detail::prepend(std::move(err), "[{}]", num);
+    dst.push_back(std::move(v));
+    num++;
+  }
+  return caf::none;
+}
+
+// Overload for maps.
+// clang-format off
```

Comment:
Can we just `clang-format` off/on just around the requires? The rest of the function is fine, right?

Reply:
I would, but clang-format 12 still acts very weirdly in the presence of anything concept related. If I do as you suggest the entire body of the function gets indented one level too deep.

Reply:
I had to check again and figured out that this problem doesn't occur for clauses that contain an ad-hoc requires expression.

Reply:
I moved the `// clang-format on` lines up where I could.